### PR TITLE
New skin visualisation

### PR DESCRIPTION
**Pull request by arrenglover**

### CODE REVIEW

**Comment by ghoshsuman** [comment]
json settings files inside `.vscode/` need to removed, and path to be added in `.gitignore`

File: src/hardwareio/binaryDumper/binary-dumper.cpp [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    yarp::os::Network yarp;
+    if(!yarp.checkNetwork(2)) {
+        std::cout << "Could not connect to YARP" << std::endl;
+        return false;
+    }
+
+    /* prepare and configure the resource finder */
+    yarp::os::ResourceFinder rf;
+    rf.setVerbose( false );
+    rf.setDefaultContext( "eventdriven" );
```
Comment (ghoshsuman) [inline src/hardwareio/binaryDumper/binary-dumper.cpp]: should it be "event-driven"?

File: lib/include/event-driven/vDraw.h [hunk 2]
```diff
@@ -1228,100 +181,15 @@ class skinDraw : public vDraw {
...
-    };
-    static const std::string drawtype;
-    virtual void draw(cv::Mat &image, const ev::vQueue &eSet, int vTime);
-    virtual std::string getDrawType();
-    virtual std::string getEventType();
-
-};
-
-
-
```
Comment (ghoshsuman) [inline lib/include/event-driven/vDraw.h]: Is this dead code being removed? Didn't find replacements being added in any other files

File: lib/include/event-driven/vDraw.h [hunk 3]
```diff
@@ -1228,100 +181,15 @@ class skinDraw : public vDraw {
...
-            cv::Point centr_all(x, y);
-            cv::circle(baseimage, centr_all, radius, black,1, CV_AA);
-        }
-    };
-
-
-
-    virtual void resetImage(cv::Mat &image){
-        baseimage.copyTo(image);
-    };
```
Comment (ghoshsuman) [inline lib/include/event-driven/vDraw.h]: Is this part of dead code for skin visualization being removed? Couldn't find replacements being added in any other files.

File: src/hardwareio/binaryDumper/binary-dumper.cpp (hunk 1 above)
Comment (ghoshsuman) [inline src/hardwareio/binaryDumper/binary-dumper.cpp]:
```suggestion
    rf.setDefaultContext( "event-driven" );
```

**Comment by arrenglover** [comment]
@ghoshsuman - I think we need to do the following changes:
1. Move the skin_draw code to a separate file - the current file becomes quite cluttered with the complexity of the skin draw code.
2. Move the code that is written in the header to the cpp files
3. review the actual implementation to make sure it is efficient
4. run some tests to make sure the visualisation is still being performed as expected. 

I'm working on points 1 & 2 - but I think we'll need to spend a little more time on 3 and ask Ali or Simon to do part 4

File: src/hardwareio/binaryDumper/binary-dumper.cpp (hunk 1 above)
Comment (arrenglover) [reply src/hardwareio/binaryDumper/binary-dumper.cpp]: this change is for the binary-dumper and therefore shouldn't be in this merge request at all. I've fixed that - but your actual comment is correct. Could you go through the entire event-driven library and fix this same issue on all the modules? thanks? then open a pull request :)

File: lib/src/vDraw_basic.cpp [hunk 4]
```diff
@@ -326,7 +213,7 @@ std::string imuDraw::getEventType()
 
 void imuDraw::draw(cv::Mat &image, const ev::vQueue &eSet, int vTime)
 {
-    auto radius = 20;
```
Comment (arrenglover) [inline lib/src/vDraw_basic.cpp]: @ghoshsuman - also note that this change is for the IMU and shouldn't be part of this pull request either. It was probably a mistake to change it but we should catch it in the code review

File: lib/include/event-driven/vDrawSkin.h [hunk 5]
```diff
@@ -0,0 +1,1159 @@
...
+        return res;
+    }
+
+    std::map<int, std::vector<double> > data;
+
+public:
+    int xoffset,yoffset;
+    int scaling =4, radius = 6;
+    int noise = 2500;
+    int max_value = 15000;//good for default flat skin patch
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: Hardcoded values not initialized in constructor. Actually tehre is no explicit constructor for the PAD class at all

File: lib/include/event-driven/vDrawSkin.h [hunk 6]
```diff
@@ -0,0 +1,1159 @@
...
+        dX[ 9]= 0; dY[ 9]=445.54;
+        dX[ 11]= 128.62; dY[ 11]=222.83;
+        dX[ 0]= 257.2; dY[ 0]=0;
+        dX[ 6]= -256.3; dY[ 6]=-147.64;
+        dX[ 1]=  385.83; dY[1]=-222.83;
+        dX[ 2]=128.62; dY[ 2]=-222.83;
+        dX[ 3]=0.0; dY[ 3]=0.0;
+        dX[ 4]=-128.62; dY[ 4]=-222.83;
+        dX[ 5]=-385.83; dY[ 5]=-222.83;
+        dX[ 7]=-257.2; dY[ 7]=0.0;
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: Again a lot of hardcoded values and the constructor doesn't take any parameter. Maybe adding a file for managing the dx and dy could be an option. The others parameters should have an initialization via constructor. nTaxels is nothing more than the length of dX and dY. very error-prone initialization

File: lib/include/event-driven/vDrawSkin.h [hunk 7]
```diff
@@ -0,0 +1,1159 @@
...
+            repr2TaxelList[mapp.front()]=vectorofIntEqualto(taxelMap,mapp.front());
+            mapp.pop_front();
+        }
+        //assign correct spacing between addresses: every 12 5 empty addresses
+        for (auto it =repr2TaxelList.begin();it!=repr2TaxelList.end();it++){
+            for (auto it2 =  it->second.begin(); it2 !=it->second.end(); it2++)
+            {
+                if(it->first==-1){
+                    s+=1;
+                    *it2 = 10000; //dummy value to avoid problems. The algorithm below works only for contiguous triangles in assigning a correct value
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: Ugly and error prone

File: lib/include/event-driven/vDrawSkin.h [hunk 8]
```diff
@@ -0,0 +1,1159 @@
...
+            xoffset = 352;
+            yoffset = 400;
+            scaling =6;
+            palmShift = shift;
+            noise = 2300;
+            max_value = 10000;
+
+            dX = new double[48];
+            dY = new double[48];
+            std::cout<< "\nInitialise hand"<<std::endl;
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: Again a lot of hardcoded values

File: lib/include/event-driven/vDrawSkin.h [hunk 9]
```diff
@@ -0,0 +1,1159 @@
...
+                        }
+                        if(s%12==0) {t+=1;s=0;}
+
+                        *it2 =128+s+ t*16;
+                        s+=1;
+
+                    }
+                    else{
+                        int triangleID = (it->first-3)/12;
+                        *it2 += triangleID *4;
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: this piece of code is very specific to the current configuration and will require additional work and headaches whenever we will change an epsilon in the taxel configuration

File: lib/include/event-driven/vDrawSkin.h [hunk 10]
```diff
@@ -0,0 +1,1159 @@
...
+                else{
+                    //yWarning("Palm");
+
+                    taxelInTriangle = (*it-128)-4*((*it-128)/16);
+
+                    triangleID = 8;
+
+                }
+                success =1;
+                break;
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: same as last comment

File: lib/include/event-driven/vDrawSkin.h [hunk 11]
```diff
@@ -0,0 +1,1159 @@
...
+            dX[1]= 15.0; dY[1]=35.0;
+            dX[2]= 41.0; dY[2]=35.0;
+            dX[3]= 30.0; dY[3]=64.0;
+            dX[4]= 11.0; dY[4]=58.0;
+            dX[5]=  0.0; dY[5]=82.0;
+            dX[6]=-11.0; dY[6]=58.0;
+            dX[7]=-30.0; dY[7]=64.0;
+            dX[8]=-41.0; dY[8]=35.0;
+            dX[9]=-15.0; dY[9]=35.0;
+            dX[10]=-15.0; dY[10]=10.0;
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: hardcoded

File: lib/include/event-driven/vDrawSkin.h [hunk 12]
```diff
@@ -0,0 +1,1159 @@
...
+            dX[36]=28;    dY[36]=-19.5;
+            dX[42]=33;    dY[42]=-19.5;
+            dX[45]=38;    dY[45]=-19.5;
+            dX[37]=28;    dY[37]=-24.5;
+            dX[38]=33;    dY[38]=-24.5;
+            dX[41]=38;    dY[41]=-24.5;
+            dX[44]=43;    dY[44]=-26;
+            dX[39]=35;    dY[39]=-29;
+            dX[40]=40;    dY[40]=-29.5;
+            dX[43]=37;    dY[43]=-32.5; //thermal pad
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: hardcoded

File: lib/include/event-driven/vDrawSkin.h [hunk 13]
```diff
@@ -0,0 +1,1159 @@
...
+            xoffset = 352;
+            yoffset = 400;
+            scaling =6;
+            palmShift = shift;
+            noise = 2300;
+            max_value = 10000;
+
+            dX = new double[48];
+            dY = new double[48];
+            std::cout<< "\nInitialise hand"<<std::endl;
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: hardcoded and not set via constructor

File: lib/include/event-driven/vDrawSkin.h [hunk 14]
```diff
@@ -0,0 +1,1159 @@
...
+                        s+=1;
+
+                    }
+                    else{
+                        int triangleID = (it->first-3)/12;
+                        *it2 += triangleID *4;
+                    }
+                }
+
+            }
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: taxel configuration specific

File: lib/include/event-driven/vDrawSkin.h [hunk 15]
```diff
@@ -0,0 +1,1159 @@
...
+                    taxelInTriangle = std :: distance(listIntriangle.begin(),it);
+                }
+                else{
+                    //yWarning("Palm");
+
+                    taxelInTriangle = (*it-128)-4*((*it-128)/16);
+
+                    triangleID = 8;
+
+                }
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: taxel configuration specific

File: lib/include/event-driven/vDrawSkin.h [hunk 16]
```diff
@@ -0,0 +1,1159 @@
...
+            dX[3]= 41.0; dY[3]=35.0;
+            dX[4]= 30.0; dY[4]=64.0;
+            dX[5]= 11.0; dY[5]=58.0;
+            dX[6]=  0.0; dY[6]=82.0;
+            dX[7]=-11.0; dY[7]=58.0;
+            dX[8]=-30.0; dY[8]=64.0;
+            dX[9]=-41.0; dY[9]=35.0;
+            dX[10]=-15.0; dY[10]=35.0;
+            dX[11]=-15.0; dY[11]=10.0;
+
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: hardcoded

File: lib/include/event-driven/vDrawSkin.h [hunk 17]
```diff
@@ -0,0 +1,1159 @@
...
+            dX[40]=33;    dY[40]=-24.5;
+            dX[38]=38;    dY[38]=-24.5;
+            dX[37]=43;    dY[37]=-26;
+            dX[41]=35;    dY[41]=-29;
+            dX[39]=40;    dY[39]=-29.5;
+            dX[43]=37;    dY[43]=-32.5; //thermal pad
+            dX[47]=33;    dY[47]=-14.5;
+            dX[46]=28;    dY[46]=-14.5;
+
+
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: hardcoded

File: lib/include/event-driven/vDrawSkin.h [hunk 18]
```diff
@@ -0,0 +1,1159 @@
...
+        {146,{9, 18}},{147,{9, 16}},{148,{10, 17}},{149,{11, 16}},{151,{10, 15}},{152,{9, 14}},{153,{8, 13}},{155,{8, 15}},{160,{10, 9}},{161,{9, 10}},
+        {162,{10, 11}},{163,{11, 10}},{164,{11, 12}},{165,{12, 13}},{167,{12, 11}},{168,{12, 9}},{169,{12, 7}},{171,{11, 8}},{176,{15, 8}},{177,{14, 7}},
+        {178,{14, 9}},{179,{15, 10}},{180,{14, 11}},{181,{14, 13}},{183,{15, 12}},{184,{16, 11}},{185,{17, 10}},{187,{16, 9}},{192,{17, 18}},{193,{18, 19}},
+        {194,{18, 17}},{195,{17, 16}},{196,{18, 15}},{197,{18, 13}},{199,{17, 14}},{200,{16, 15}},{201,{15, 16}},{203,{16, 17}},{208,{16, 23}},{209,{17, 22}},
+        {210,{16, 21}},{211,{15, 22}},{212,{15, 20}},{213,{14, 19}},{215,{14, 21}},{216,{14, 23}},{217,{14, 25}},{219,{15, 24}},{224,{16, 3}},{225,{15, 4}},
+        {226,{16, 5}},{227,{17, 4}},{228,{17, 6}},{229,{18, 7}},{231,{18, 5}},{232,{18, 3}},{233,{18, 1}},{235,{17, 2}},{240,{21, 2}},{241,{20, 1}},
+        {242,{20, 3}},{243,{21, 4}},{244,{20, 5}},{245,{20, 7}},{247,{21, 6}},{248,{22, 5}},{249,{23, 4}},{251,{22, 3}}};
+
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: hardcoded. again i think we need a file for these type of parameters

File: lib/include/event-driven/vDrawSkin.h [hunk 19]
```diff
@@ -0,0 +1,1159 @@
...
+        loadMap(){
+
+            noise = 2500;
+            max_value = 15000;
+
+            xoffset = 80;
+            //yoffset = 120;
+            yoffset = 680;
+            scaling=20;
+            radius =10;
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: hardcoded and not set in constructor

File: lib/include/event-driven/vDrawSkin.h [hunk 20]
```diff
@@ -0,0 +1,1159 @@
...
+                    max_value = handL.max_value;
+                }
+                else {
+                    yWarning("Unable to build the taxel map from file");
+                    pos = default_pos;
+                }
+            }
+
+        };
+
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: A lot of duplicated code that would be hard to maintain in the future. Also is not clear what this "check" variable is. It seems like it is used to move to the next part of the initialization. Probably could be replaced by return statements if the init function cannot proceed for some reasons

File: lib/include/event-driven/vDrawSkin.h [hunk 21]
```diff
@@ -0,0 +1,1159 @@
...
+        yInfo("scaling = %d", tmap.scaling );
+        yInfo("xoffset = %d", tmap.xoffset );
+        yInfo("yoffset =%d", tmap.yoffset );
+
+        for (auto it = tmap.pos.begin(); it != tmap.pos.end(); it++)
+        {
+            x = tmap.xoffset + tmap.scaling* std :: get<0>(it->second);
+            y =  tmap.yoffset +tmap.scaling* std :: get<1>(it->second);
+            cv::Point centr_all(x, y);
+            cv::circle(baseimage, centr_all, radius, black,1, CV_AA);
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: This is the only part that really changes from the previous implementations of this class

File: lib/include/event-driven/vDrawSkin.h [hunk 22]
```diff
@@ -0,0 +1,1159 @@
...
+        int x;
+        int y;
+
+
+        for (auto it = tmap.pos.begin(); it != tmap.pos.end(); it++)
+        {
+            x = tmap.xoffset + tmap.scaling* std :: get<0>(it->second);
+            y =  tmap.yoffset +tmap.scaling* std :: get<1>(it->second);
+            cv::Point centr_all(x, y);
+            cv::circle(baseimage, centr_all, radius, black,1, CV_AA);
```
Comment (Iaxama) [inline lib/include/event-driven/vDrawSkin.h]: This is the only part that really changes from the previous implementations of this class

File: lib/include/event-driven/vDrawSkin.h [hunk 23]
```diff
@@ -0,0 +1,1159 @@
+/*
+ *   Copyright (C) 2017 Event-driven Perception for Robotics
```
Comment (aikolina) [inline lib/include/event-driven/vDrawSkin.h]: Copyright year to be updated.
```suggestion
 *   Copyright (C) 2020 Event-driven Perception for Robotics
```

File: lib/src/vDraw_skin.cpp [hunk 24]
```diff
@@ -0,0 +1,349 @@
+/*
+ *   Copyright (C) 2017 Event-driven Perception for Robotics
```
Comment (aikolina) [inline lib/src/vDraw_skin.cpp]: Copyright year to be updated.
```suggestion
 *   Copyright (C) 2020 Event-driven Perception for Robotics
```

File: src/processing/vFramer/src/vFramerLite.cpp [hunk 25]
```diff
@@ -67,6 +68,12 @@ vDraw * createDrawer(std::string tag)
         return new skinDraw();
     if(tag == skinsampleDraw::drawtype)
         return new skinsampleDraw();
+    if(tag == isoDrawSkin::drawtype)
```
Comment (aikolina) [inline src/processing/vFramer/src/vFramerLite.cpp]: Wouldn't it be more convenient to change these multiple `if` statements into a `switch case` statement?
It would improve the readability and it should be faster (not always true).

File: lib/include/event-driven/vDrawSkin.h (hunk 5 above)
Comment (aikolina) [inline lib/include/event-driven/vDrawSkin.h]: It should be sufficient to initialize the values using a list initialization.
```suggestion
    int scaling{4};
    int radius{6};
    int noise{2500};
    int max_value{15000};//good for default flat skin patch
```

File: lib/include/event-driven/vDrawSkin.h [hunk 26]
```diff
@@ -0,0 +1,1159 @@
...
+                res.push_back(i);
+            }
+        }
+        return res;
+    }
+
+    std::map<int, std::vector<double> > data;
+
+public:
+    int xoffset,yoffset;
```
Comment (aikolina) [inline lib/include/event-driven/vDrawSkin.h]: What is the default value for `xoffset` and `yoffset`?

File: lib/include/event-driven/vDrawSkin.h [hunk 27]
```diff
@@ -0,0 +1,1159 @@
...
+            for (auto it =repr2TaxelList.begin();it!=repr2TaxelList.end();it++){
+                for (auto it2 =  it->second.begin(); it2 !=it->second.end(); it2++){
+                    if(it->first ==-1)*it2=10000;//dummy value ;
+
+                    if(it->first>60){
+                        //if(s==5) s+=1;/ skip 6th
+
+                        if (t==3){
+                            if(s==7)s+=1;
+                            }
```
Comment (aikolina) [inline lib/include/event-driven/vDrawSkin.h]: indentation
```suggestion
                        }
```

File: lib/include/event-driven/vDrawSkin.h [hunk 28]
```diff
@@ -0,0 +1,1159 @@
...
+
+class PAD{
+protected :
+
+    const double DEG2RAD=M_PI/180.0;
+    int nTaxels;
+
+    double *dX,*dY;
+
+    inline std::list<unsigned int> vectorofIntEqualto(const std::vector<int> vec, const int val)
```
Comment (aikolina) [inline lib/include/event-driven/vDrawSkin.h]: IMO, the name of the function is not representative of its actual role because I would think: "equal to what?"
Possible alternatives (I couldn't think of anything better):
- `vectorOfIntEqualToInt()` 
- `compareVectorOfIntToInt()`

File: lib/include/event-driven/vDrawSkin.h [hunk 29]
```diff
@@ -0,0 +1,1159 @@
...
+
+        if (triangleID==-1){
+            for(auto mapit=repr2TaxelList.begin();mapit!=repr2TaxelList.end();mapit++){
+            auto listIntriangle= mapit->second;
+            auto it = std::find(listIntriangle.begin(), listIntriangle.end(), index-1);
+            if(it != listIntriangle.end()){
+                triangleID = mapit->first;
+                break;
+                }
+            }
```
Comment (aikolina) [inline lib/include/event-driven/vDrawSkin.h]: indentation
```suggestion
            for(auto mapit=repr2TaxelList.begin();mapit!=repr2TaxelList.end();mapit++){
                auto listIntriangle= mapit->second;
                auto it = std::find(listIntriangle.begin(), listIntriangle.end(), index-1);
                if(it != listIntriangle.end()){
                    triangleID = mapit->first;
                    break;
                }
            }
```

File: lib/include/event-driven/vDrawSkin.h [hunk 30]
```diff
@@ -0,0 +1,1159 @@
...
+    int xoffset,yoffset;
+    int scaling =4, radius = 6;
+    int noise = 2500;
+    int max_value = 15000;//good for default flat skin patch
+
+    std::map<int, std::list<unsigned int> > repr2TaxelList;
+    virtual void initRepresentativeTaxels(std::vector<int> taxelMap){};
+    virtual int* getID(const int index) const {};
+    virtual std::tuple<double,double> makeMap( int ID, int taxelIntriangle) const {};
+    void get_data(int id,double x,double y, double orientation,double gain, int mirror,int layoutNum){
```
Comment (aikolina) [inline lib/include/event-driven/vDrawSkin.h]: `layoutNum` is not used.

File: lib/include/event-driven/vDrawSkin.h [hunk 31]
```diff
@@ -0,0 +1,1159 @@
...
+
+                        int id=sensorConfig.get(1).asInt();
+                        double x=sensorConfig.get(2).asDouble();
+                        double y=sensorConfig.get(3).asDouble();
+                        double orientation=sensorConfig.get(4).asDouble();
+                        double gain=sensorConfig.get(5).asDouble();
+                        int    mirror=sensorConfig.get(6).asInt();
+                        int    layoutNum=sensorConfig.get(7).asInt();
+
+                        skin.get_data(id,x,y,orientation,gain,mirror,layoutNum);
```
Comment (aikolina) [inline lib/include/event-driven/vDrawSkin.h]: `layoutNum` is not used.
```suggestion
                        skin.get_data(id,x,y,orientation,gain,mirror);
```

File: lib/include/event-driven/vDrawSkin.h [hunk 32]
```diff
@@ -0,0 +1,1159 @@
...
+
+                        int id=sensorConfig.get(1).asInt();
+                        double x=sensorConfig.get(2).asDouble();
+                        double y=sensorConfig.get(3).asDouble();
+                        double orientation=sensorConfig.get(4).asDouble();
+                        double gain=sensorConfig.get(5).asDouble();
+                        int    mirror=sensorConfig.get(6).asInt();
+                        int    layoutNum=sensorConfig.get(7).asInt();
+
+                        handR.get_data(id,x,y,orientation,gain,mirror,layoutNum);
```
Comment (aikolina) [inline lib/include/event-driven/vDrawSkin.h]: `layoutNum` is not used.
```suggestion
                        handR.get_data(id,x,y,orientation,gain,mirror);
```

File: lib/include/event-driven/vDrawSkin.h [hunk 33]
```diff
@@ -0,0 +1,1159 @@
...
+
+                        int id=sensorConfig.get(1).asInt();
+                        double x=sensorConfig.get(2).asDouble();
+                        double y=sensorConfig.get(3).asDouble();
+                        double orientation=sensorConfig.get(4).asDouble();
+                        double gain=sensorConfig.get(5).asDouble();
+                        int    mirror=sensorConfig.get(6).asInt();
+                        int    layoutNum=sensorConfig.get(7).asInt();
+
+                        handL.get_data(id,x,y,orientation,gain,mirror,layoutNum);
```
Comment (aikolina) [inline lib/include/event-driven/vDrawSkin.h]: `layoutNum` is not used.
```suggestion
                        handL.get_data(id,x,y,orientation,gain,mirror);
```

File: lib/include/event-driven/vDrawSkin.h [hunk 34]
```diff
@@ -0,0 +1,1159 @@
...
+
+        int ID = getID(index)[0];
+        int taxelIntriangle =getID(index)[1];
+
+        std::vector<double> info = data.find(ID)->second;
+
+        double cx = info[0];
+        double cy = info[1];
+        double th = info[2];
+        double gain = info[3];
```
Comment (aikolina) [inline lib/include/event-driven/vDrawSkin.h]: `gain` is not used.
```suggestion

```

File: lib/CMakeLists.txt [hunk 35]
```diff
@@ -52,7 +53,8 @@ set(folder_header
 
 if(OpenCV_FOUND)
     list(APPEND folder_header include/event-driven/vIPT.h
-                              include/event-driven/vDraw.h)
+                              include/event-driven/vDraw.h
+                              include/event-driven/vDrawSkin.h)
 endif()
```
[the remark this answers is not in the document]
Comment (arrenglover) [reply lib/CMakeLists.txt]: sorry massi I'm not sure what the change is you want to make here?

File: lib/src/vDraw_skin.cpp [hunk 36]
```diff
@@ -0,0 +1,349 @@
...
+                cv::circle(image, centr, radius, red, 1, CV_AA);
+            }
+        }
+        else{
+            //yWarning() << "\n Index " << index <<"not mapped! \n";
+            //std :: cout << "\n Index " << index <<"not mapped! \n";
+        }
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (arrenglover) [reply lib/src/vDraw_skin.cpp]: Is there a change to be made?

[the remark this answers is not in the document]
Comment (Iaxama) [reply lib/src/vDraw_skin.cpp]: I think this part is good

File: lib/include/event-driven/vDrawSkin.h (hunk 5 above)
Comment (arrenglover) [reply lib/include/event-driven/vDrawSkin.h]: I just added some default values{0}. No need for the default constructor. I fixed some virtual class initialisation too.

File: lib/include/event-driven/vDrawSkin.h (hunk 18 above)
Comment (arrenglover) [reply lib/include/event-driven/vDrawSkin.h]: yes the plan was/is to read these values from the iCubSkinGIU - for now they will remain hardcoded for the skin patch we have and further development will need to be made

File: lib/include/event-driven/vDrawSkin.h (hunk 6 above)
Comment (arrenglover) [reply lib/include/event-driven/vDrawSkin.h]: agreed, this should be fixed if/when loading from the file gets done. I added a deconstructor which deletes these correctly.

File: lib/include/event-driven/vDrawSkin.h (hunk 7 above)
Comment (arrenglover) [reply lib/include/event-driven/vDrawSkin.h]: yeah agreed, but there is some not easy to understand interpretation of the data we have. I don't want to play with this now.

File: lib/include/event-driven/vDrawSkin.h (hunk 8 above)
Comment (arrenglover) [reply lib/include/event-driven/vDrawSkin.h]: agreed. for now we will leave them as is

File: lib/include/event-driven/vDrawSkin.h (hunk 9 above)
Comment (arrenglover) [reply lib/include/event-driven/vDrawSkin.h]: agreed - deal with it when someone develops loading from config file

File: lib/include/event-driven/vDrawSkin.h (hunk 20 above)
Comment (arrenglover) [reply lib/include/event-driven/vDrawSkin.h]: agreed, for now it "works" but should be made more understandable later

File: lib/include/event-driven/vDrawSkin.h (hunk 26 above)
Comment (arrenglover) [reply lib/include/event-driven/vDrawSkin.h]: I added a default - hopefully it is correct.

File: lib/include/event-driven/vDrawSkin.h (hunk 27 above)
Comment (arrenglover) [reply lib/include/event-driven/vDrawSkin.h]: I ran auto-indent on it.

File: src/processing/vFramer/src/vFramerLite.cpp (hunk 25 above)
Comment (arrenglover) [reply src/processing/vFramer/src/vFramerLite.cpp]: you can't switch on a std::string

File: lib/include/event-driven/vDrawSkin.h (hunk 28 above)
Comment (arrenglover) [reply lib/include/event-driven/vDrawSkin.h]: agreed, there are many changes that should be made to improve these things so I'll leave this one for the moment

File: lib/include/event-driven/vDrawSkin.h (hunk 29 above)
Comment (arrenglover) [reply lib/include/event-driven/vDrawSkin.h]: I ran an auto indent

File: lib/include/event-driven/vDrawSkin.h (hunk 31 above)
Comment (arrenglover) [reply lib/include/event-driven/vDrawSkin.h]: agreed - at this point I'm not sure if it will break something else so I'll leave it as is.

File: lib/include/event-driven/vDrawSkin.h (hunk 30 above)
Comment (arrenglover) [reply lib/include/event-driven/vDrawSkin.h]: agreed - I'm not sure it will break something else so I'll leave it.

**Comment by arrenglover** [comment]
Thanks @Iaxama @aikolina @ghoshsuman for checking this code. I completely agree with your concerns of this code. I've done a few more small improvements, but acutally resolving the concerns requires someone dedicated on this task. I think it's best to move forward in this way:
0. I've already separated the skin visualisation code into it's own file so it is contained and we know all the files/functions concerned.
1. merge this code in to master
2. have @dabbous94 test it as soon as possible - it should work and give us the nice visualisation for skin.
3. as Ali and Simon continue to work on the skin have them solve these issues and improve it as it is needed

I really prefer not to leave it in a pull request/branch as I know it's going to create more issues with getting it completed than just merging it now will be. We want to use this visualisation on the robot immediately.

**Comment by aikolina** [comment]
@arrenglover Since we found a long list of TODOs, I think we need to keep track of them somehow. I suggest to open a new issue and add a tag `TODO` within the code where is needed. I can take care of that!

**Comment by Iaxama** [comment]
I agree with both of you Let's merge it into master and start use this, but also let's add a way to track the fixes to be done. One way would be to have one or several issues or to add TODOs directly in the code. Maybe having both is a good idea

**Comment by arrenglover** [comment]
I'm happy to try to remember the changes to make in some way. I think that basically the whole thing needs an overhaul anyway and once we start getting into it basically everything will change, so pointing out specific line numbers etc. might be a waste. Opening an issues that we still need to do the loading of the taxel positions from files would be the minimum though.

**Comment by aikolina** [comment]
@arrenglover Just to clarify because I'm not sure I conveyed my idea properly: I wasn't suggesting to point out the line numbers in the issue but to add a `TODO: reasonable comment` within the code itself where needed. I thought it could be handy for whoever has to do the clean-up later since many IDE highlight `TODO` keywords.

> Opening an issues that we still need to do the loading of the taxel positions from files would be the minimum though.

I'm going to do it.
